orttraining/orttraining/training_ops/cpu/tensor/split.cc

```diff
@@ -0,0 +1,153 @@
+                                   : gsl::narrow<int>(input_shape.SizeFromDimension(axis + 1));
+
+  std::vector<int64_t> split_sizes_values(split_sizes); 
+  split_sizes.clear();
+  int64_t split_size_sum = std::accumulate(split_sizes_values.cbegin(), split_sizes_values.cend(), 0LL);
+
+  if (split_sizes_values.empty()) {
+    // equal split based on number of outputs
+    if (split_dim_size % static_cast<size_t>(num_outputs) != 0) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input cannot be split evenly on selected axis. Input shape=", input_shape,
+                             " Axis=", axis_, " NumOutputs=", num_outputs);
+    }
+
+    // populate split_sizes with the same size for each output
+    split_sizes = std::vector<int64_t>(static_cast<size_t>(num_outputs), split_dim_size / num_outputs);
+  } else {
+    if (split_sizes_values.size() != static_cast<size_t>(num_outputs) || split_size_sum != split_dim_size)
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                             "Cannot split using values in 'split' input. Axis=", axis_,
+                             " Input shape=", input_shape,
+                             " NumOutputs=", num_outputs,
+                             " Num entries in 'split' (must equal number of outputs) was ", split_sizes_values.size(),
+                             " Sum of sizes in 'split' (must equal size of selected axis) was ", split_size_sum);
+
+    split_sizes = split_sizes_values;
+  }
+
+  return Status::OK();
+}
+
+Status SplitTraining::Compute(OpKernelContext* context) const {
+  const Tensor& input = *context->Input<Tensor>(0);
+
+  Status status;
+
+  if (input.IsDataType<float>())
+    status = ComputeImpl<float>(*context, input);
+  else if (input.IsDataType<int32_t>())
+    status = ComputeImpl<int32_t>(*context, input);
+  else if (input.IsDataType<int64_t>())
+    status = ComputeImpl<int64_t>(*context, input);
+  else if (input.IsDataTypeString())
+    status = ComputeImpl<std::string>(*context, input);
+  else
+    ORT_THROW("Split operator does not support ", input.DataType(), " yet");
+
+  return status;
+}
+
+template <typename T>
+inline void copy_data(const T* src, T* dst, size_t count) {
+  memcpy(dst, src, count * sizeof(T));
+}
+
+template <>
+inline void copy_data<std::string>(const std::string* src, std::string* dst, size_t count) {
+  const std::string* end = src + count;
+  std::copy(src, end, dst);
+}
+
+template <typename T>
+Status SplitTraining::ComputeImpl(OpKernelContext& context, const Tensor& input) const {
+  auto& input_shape = input.Shape();
+  auto num_outputs = context.OutputCount();
+  int64_t axis = axis_;
+  int before_dims = 0;
+  int after_dims_including_split_axis = 0;
+  int after_dims_excluding_split = 0;
+
+  //override the attribute value with the input value for split_split
+  const Tensor* split_tensor = context.Input<Tensor>(1);
+  ORT_ENFORCE(split_tensor != nullptr, "Split input is null");
```

Comment:
should this be triggered for the test cases where split input is missing?

Reply:
same for cuda kernel

Reply:
in fact, it will fail on the schema check, won't even get here. I'll remove the checks

---
In reply to: [459791279](https://github.com/microsoft/onnxruntime/pull/4597#discussion_r459791279) [](ancestors = 459791279)